rust/extractor/src/crate_graph.rs

```diff
@@ -0,0 +1,655 @@
+use crate::{
+    generated,
+    trap::{self, TrapFile},
+};
+use chalk_ir::FloatTy;
+use chalk_ir::IntTy;
+use chalk_ir::Scalar;
+use chalk_ir::UintTy;
+use itertools::Itertools;
+use ra_ap_base_db::CrateGraph;
+use ra_ap_base_db::CrateId;
+use ra_ap_base_db::SourceDatabase;
+use ra_ap_cfg::CfgAtom;
+use ra_ap_hir::{db::DefDatabase, VariantId};
+use ra_ap_hir::{db::HirDatabase, DefMap, ModuleDefId};
+use ra_ap_hir_def::HasModule;
+use ra_ap_hir_def::{data::adt::VariantData, AssocItemId, LocalModuleId};
+use ra_ap_hir_ty::Interner;
+use ra_ap_hir_ty::TraitRefExt;
+use ra_ap_hir_ty::Ty;
+use ra_ap_hir_ty::TyExt;
+use ra_ap_hir_ty::WhereClause;
+use ra_ap_hir_ty::{Binders, FnPointer};
+use ra_ap_ide_db::RootDatabase;
+use ra_ap_vfs::{Vfs, VfsPath};
+use std::hash::Hash;
+use std::hash::Hasher;
+use std::{cmp::Ordering, collections::HashMap, path::PathBuf};
+
+pub fn extract_crate_graph(trap_provider: &trap::TrapFileProvider, db: &RootDatabase, vfs: &Vfs) {
+    let crate_graph = db.crate_graph();
+
+    // According to the documentation of `CrateGraph`:
+    // Each crate is defined by the `FileId` of its root module, the set of enabled
+    // `cfg` flags and the set of dependencies.
+    let mut crate_id_map = HashMap::<CrateId, (PathBuf, u64)>::new();
+    for krate_id in crate_graph.crates_in_topological_order() {
+        let krate = &crate_graph[krate_id];
+        let root_module_file: &VfsPath = vfs.file_path(krate.root_file_id);
+        if let Some(root_module_file) = root_module_file
+            .as_path()
+            .map(|p| std::fs::canonicalize(p).unwrap_or(p.into()))
+        {
+            let mut hasher = std::hash::DefaultHasher::new();
+            krate
+                .cfg_options
+                .as_ref()
+                .into_iter()
+                .sorted_by(cmp_flag)
+                .for_each(|x| format!("{x}").hash(&mut hasher));
+
+            krate
+                .dependencies
+                .iter()
+                .flat_map(|d| crate_id_map.get(&d.crate_id))
+                .sorted()
+                .for_each(|x| x.hash(&mut hasher));
+            let hash = hasher.finish();
+            crate_id_map.insert(krate_id, (root_module_file, hash));
+        }
+    }
+    for krate_id in crate_graph.iter() {
+        if let Some((root_module_file, hash)) = crate_id_map.get(&krate_id) {
+            let path = root_module_file.join(format!("{hash:0>16x}"));
+            let mut trap = trap_provider.create("crates", path.as_path());
+            if trap.path.exists() {
+                continue;
+            }
```

Comment:
maybe we need a comment here that this is a potential race condition if we ever start doing this in parallel...